src/Database/Oracle/Simple/FromField.hs

```diff
@@ -122,17 +123,38 @@ getWord64 = dpiData_getUint64
 getBool :: ReadDPIBuffer Bool
 getBool ptr = (== 1) <$> dpiData_getBool ptr
 
--- | Get Text from the data buffer
+-- | Get Text from the data buffer.
+-- Supports ASCII, UTF-8 and UTF-16 big- and little-endian encodings.
+-- Throws 'FieldParseError' if any other encoding is encountered.
 getText :: ReadDPIBuffer Text
-getText = fmap pack <$> getString
+getText = buildText <=< peek <=< dpiData_getBytes
+ where
+  buildText DPIBytes{..} = do
+    gotBytes <- BS.packCStringLen (dpiBytesPtr, fromIntegral dpiBytesLength)
+    encoding <- peekCString dpiBytesEncoding
+    decodeFn <- case encoding of
+      "ASCII" -> pure decodeASCII
+      "UTF-8" -> pure decodeUtf8
+      "UTF-16BE" -> pure decodeUtf16BE
+      "UTF-16LE" -> pure decodeUtf16LE
+      otherEnc -> throwIO $ ByteDecodeError otherEnc
+    pure $ decodeFn gotBytes
```

Comment:
We might want to try and catch `decodeFn` since it /can/ throw an exception

```haskell
evaluate (decodeFn gotBytes) `catch` (\(e :: SomeException) -> 
  throwIO (ByteDecodeError (show e))
```

Reply:
I thought we could trust the encoding string Oracle sends us, but you're right, the function itself still can throw an exception. I shall add a catch here, or I'll use the safe versions of the decode functions instead.